tests/lax_numpy_test.py

```diff
@@ -2090,6 +2090,24 @@ def testTrilIndicesFrom(self, shape, dtype, k):
     args_maker = lambda: [rng(shape, dtype), k]
     self._CheckAgainstNumpy(np_fun, jnp_fun, args_maker)
 
+  @jtu.sample_product(
+    dtype=default_dtypes,
+    shape=[(0, 0), (0, 1), (1, 0), (1, 1), (2, 0), (2, 1), (2, 2), (1, 2), (0, 2), (2, 3), (2, 2, 2), (2, 2, 2, 2)],
+    scalar_val=[True, False],
+  )
+  def testFillDiagonal(self, dtype, shape, scalar_val):
+    rng = jtu.rand_default(self.rng())
+
+    def np_fun(a, val):
+      a_copy = a.copy()
+      np.fill_diagonal(a_copy, val)
+      return a_copy
+
+    jnp_fun = partial(jnp.fill_diagonal, inplace=False)
+    args_maker = lambda : [rng(shape, dtype), rng(() if scalar_val else min(shape), dtype)]
```

Comment:
`np.fill_diagonal` supports `val` with shape not equal to the size of the diagonal. I'm not sure how important that is to support in JAX though...

Reply:
@jakevdp The [implementation](https://github.com/numpy/numpy/blob/v1.26.0/numpy/lib/index_tricks.py#L916) of [`numpy.fill_diagonal`](https://numpy.org/doc/stable/reference/generated/numpy.fill_diagonal.html) does the following:
```python3
a.flat[:end:step] = val
```
which uses [`numpy.ndarray.flat`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.flat.html). What would be the JAX equivalent of this?

Reply:
I don’t think there’s a direct Jax equivalent to the semantics of `np.ndarray.flat.__setitem__`

Reply:
@jakevdp In that case, would it be acceptable to just implement `fill_diagonal` as
```python3
def fill_diagonal(a: ArrayLike, val: ArrayLike, wrap: bool = False, *, inplace: bool = True) -> Array:
  if inplace:
    raise NotImplementedError("JAX arrays are immutable, must use inplace=False")
  util.check_arraylike("fill_diagonal", a, val)
  a = asarray(a)
  if a.ndim < 2:
    raise ValueError("array must be at least 2-d")
  end = None
  if a.ndim == 2:
    step = a.shape[1] + 1
    if not wrap:
      end = a.shape[1] * a.shape[1]
  else:
    if not all(n == a.shape[0] for n in a.shape[1:]):
      raise ValueError("All dimensions of input must be of equal length")
    step = 1 + np.cumprod(a.shape[:-1]).sum()
  return a.ravel().at[:end:step].set(val).reshape(a.shape)
```
?

Reply:
I think we should avoid flattening and unflattening, because that tends not to play well with sharded data. In any case, though, I don't think this implementation has the same semantics as the original function. From the [docs](https://numpy.org/doc/stable/reference/generated/numpy.fill_diagonal.html):
```
val : scalar or array_like
  Value(s) to write on the diagonal. If `val` is scalar, the value is
  written along the diagonal. If array-like, the flattened `val` is
  written along the diagonal, repeating if necessary to fill all
  diagonal entries.
```
To replicate those semantics, I think you could change the return statement to something like this:
```python
return asarray(a).at[idx].set(_tile_to_size(val.ravel(), diag_size))
```